File: testflight/testflight_deployment.py
"""
iOS CI/CD Pipeline Script for TestFlight Deployment
Handles building, testing, and uploading iOS apps to TestFlight
"""

import os
import sys
import subprocess
import json
import argparse
import time
from pathlib import Path
from typing import Dict, List, Optional

class IOSPipeline:
    def __init__(self, config: Dict):
        self.config = config
        self.project_path = config.get('project_path', '.')
        self.scheme = config.get('scheme')
        self.workspace = config.get('workspace')
        self.project = config.get('project')
        self.configuration = config.get('configuration', 'Release')
        self.team_id = config.get('team_id')
        self.bundle_id = config.get('bundle_id')
        self.app_store_connect_api_key_id = config.get('app_store_connect_api_key_id')
        self.app_store_connect_api_issuer_id = config.get('app_store_connect_api_issuer_id')
        self.app_store_connect_api_key_path = config.get('app_store_connect_api_key_path')

    def run_command(self, command: List[str], cwd: Optional[str] = None) -> subprocess.CompletedProcess:
        """Execute a shell command and return the result"""
        print(f"🔧 Running: {' '.join(command)}")
        try:
            result = subprocess.run(
                command,
                cwd=cwd or self.project_path,
                capture_output=True,
                text=True,
                check=True
            )
            if result.stdout:
                print(f"✅ Output: {result.stdout}")
            return result
        except subprocess.CalledProcessError as e:
            print(f"❌ Error running command: {e}")
            print(f"❌ Stderr: {e.stderr}")
            print(f"❌ Stdout: {e.stdout}")
            raise

    def check_prerequisites(self) -> bool:
        """Check if all required tools and configurations are available"""
        print("🔍 Checking prerequisites...")

        # Check if Xcode is installed
        try:
            self.run_command(['xcodebuild', '-version'])
        except subprocess.CalledProcessError:
            print("❌ Xcode is not installed or not in PATH")
            return False

        # Check if xcpretty is available (optional but recommended)
        try:
            subprocess.run(['xcpretty', '--version'], capture_output=True, check=True)
            self.has_xcpretty = True
        except (subprocess.CalledProcessError, FileNotFoundError):
            print("⚠️  xcpretty not found. Install with: gem install xcpretty")
            self.has_xcpretty = False

        # Check required configuration
        required_fields = ['scheme']
        if not (self.workspace or self.project):
            print("❌ Either workspace or project must be specified")
            return False

        for field in required_fields:
            if not getattr(self, field):
                print(f"❌ Missing required configuration: {field}")
                return False

        print("✅ Prerequisites check passed")
        return True

    def clean_project(self) -> bool:
        """Clean the project"""
        print("🧹 Cleaning project...")
        try:
            build_target = ['-workspace', self.workspace] if self.workspace else ['-project', self.project]
            command = ['xcodebuild', 'clean'] + build_target + ['-scheme', self.scheme]
            self.run_command(command)
            print("✅ Project cleaned successfully")
            return True
        except subprocess.CalledProcessError:
            print("❌ Failed to clean project")
            return False

    def install_dependencies(self) -> bool:
        """Install CocoaPods dependencies if Podfile exists"""
        podfile_path = Path(self.project_path) / 'Podfile'
        if podfile_path.exists():
            print("📦 Installing CocoaPods dependencies...")
            try:
                self.run_command(['pod', 'install'])
                print("✅ CocoaPods dependencies installed")
                return True
            except subprocess.CalledProcessError:
                print("❌ Failed to install CocoaPods dependencies")
                return False
        else:
            print("ℹ️  No Podfile found, skipping dependency installation")
            return True

    def run_tests(self) -> bool:
        """Run unit and UI tests"""
        print("🧪 Running tests...")
        try:
            build_target = ['-workspace', self.workspace] if self.workspace else ['-project', self.project]
            command = [
                          'xcodebuild', 'test',
                          '-scheme', self.scheme,
                          '-destination', 'platform=iOS Simulator,name=iPhone 15,OS=latest'
                      ] + build_target

            if self.has_xcpretty:
                # Pipe output through xcpretty for better formatting
                xcodebuild = subprocess.Popen(
                    command,
                    cwd=self.project_path,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    text=True
                )
                xcpretty = subprocess.Popen(
                    ['xcpretty', '--test', '--color'],
                    stdin=xcodebuild.stdout,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    text=True
                )
                xcodebuild.stdout.close()
                output, error = xcpretty.communicate()

                if xcodebuild.wait() != 0 or xcpretty.returncode != 0:
                    raise subprocess.CalledProcessError(1, command)
                print(output)
            else:
                self.run_command(command)

            print("✅ All tests passed")
            return True
        except subprocess.CalledProcessError:
            print("❌ Tests failed")
            return False

    def build_archive(self) -> Optional[str]:
        """Build and archive the app"""
        print("🔨 Building and archiving app...")

        archive_path = Path(self.project_path) / 'build' / f'{self.scheme}.xcarchive'
        archive_path.parent.mkdir(exist_ok=True)

        try:
            build_target = ['-workspace', self.workspace] if self.workspace else ['-project', self.project]
            command = [
                          'xcodebuild', 'archive',
                          '-scheme', self.scheme,
                          '-configuration', self.configuration,
                          '-archivePath', str(archive_path),
                          '-destination', 'generic/platform=iOS'
                      ] + build_target

            if self.team_id:
                command.extend(['-allowProvisioningUpdates'])
                command.extend(['DEVELOPMENT_TEAM=' + self.team_id])

            self.run_command(command)
            print(f"✅ Archive created at: {archive_path}")
            return str(archive_path)
        except subprocess.CalledProcessError:
            print("❌ Failed to build archive")
            return None

    def export_ipa(self, archive_path: str) -> Optional[str]:
        """Export IPA from archive"""
        print("📦 Exporting IPA...")

        export_path = Path(self.project_path) / 'build' / 'export'
        export_path.mkdir(exist_ok=True)

        # Create export options plist
        export_options = {
            'method': 'app-store',
            'uploadBitcode': False,
            'uploadSymbols': True,
            'compileBitcode': False
        }

        if self.team_id:
            export_options['teamID'] = self.team_id

        export_plist_path = export_path / 'ExportOptions.plist'

        # Convert dict to plist format
        import plistlib
        with open(export_plist_path, 'wb') as f:
            plistlib.dump(export_options, f)

        try:
            command = [
                'xcodebuild', '-exportArchive',
                '-archivePath', archive_path,
                '-exportPath', str(export_path),
                '-exportOptionsPlist', str(export_plist_path)
            ]

            self.run_command(command)

            # Find the exported IPA file
            ipa_files = list(export_path.glob('*.ipa'))
            if ipa_files:
                ipa_path = ipa_files[0]
                print(f"✅ IPA exported to: {ipa_path}")
                return str(ipa_path)
            else:
                print("❌ No IPA file found after export")
                return None
        except subprocess.CalledProcessError:
            print("❌ Failed to export IPA")
            return None

    def upload_to_testflight(self, ipa_path: str) -> bool:
        """Upload IPA to TestFlight using App Store Connect API"""
        print("🚀 Uploading to TestFlight...")

        try:
            command = ['xcrun', 'altool', '--upload-app', '--file', ipa_path, '--type', 'ios']

            # Use App Store Connect API key if available
            if all([self.app_store_connect_api_key_id,
                    self.app_store_connect_api_issuer_id,
                    self.app_store_connect_api_key_path]):
                command.extend([
                    '--apiKey', self.app_store_connect_api_key_id,
                    '--apiIssuer', self.app_store_connect_api_issuer_id
                ])

                # Set the API key file path as environment variable
                env = os.environ.copy()
                env['API_PRIVATE_KEYS_DIR'] = str(Path(self.app_store_connect_api_key_path).parent)
            else:
                print("⚠️  App Store Connect API credentials not provided")
                print("   You'll need to authenticate manually or provide API credentials")
                env = None

            result = subprocess.run(
                command,
                cwd=self.project_path,
                capture_output=True,
                text=True,
                env=env
            )

            if result.returncode == 0:
                print("✅ Successfully uploaded to TestFlight")
                print("ℹ️  Processing may take a few minutes in App Store Connect")
                return True
            else:
                print(f"❌ Upload failed: {result.stderr}")
                return False

        except Exception as e:
            print(f"❌ Upload error: {e}")
            return False

    def increment_build_number(self) -> bool:
        """Increment the build number in Info.plist"""
        print("🔢 Incrementing build number...")
        try:
            command = [
                'xcrun', 'agvtool', 'next-version', '-all'
            ]
            self.run_command(command)
            print("✅ Build number incremented")
            return True
        except subprocess.CalledProcessError:
            print("❌ Failed to increment build number")
            return False

    def run_pipeline(self, skip_tests: bool = False, skip_upload: bool = False) -> bool:
        """Run the complete CI/CD pipeline"""
        print("🚀 Starting iOS CI/CD Pipeline")
        print("=" * 50)

        # Check prerequisites
        if not self.check_prerequisites():
            return False

        # Install dependencies
        if not self.install_dependencies():
            return False

        # Clean project
        if not self.clean_project():
            return False

        # Increment build number
        if not self.increment_build_number():
            print("⚠️  Failed to increment build number, continuing...")

        # Run tests
        if not skip_tests:
            if not self.run_tests():
                return False
        else:
            print("⏭️  Skipping tests")

        # Build archive
        archive_path = self.build_archive()
        if not archive_path:
            return False

        # Export IPA
        ipa_path = self.export_ipa(archive_path)
        if not ipa_path:
            return False

        # Upload to TestFlight
        if not skip_upload:
            if not self.upload_to_testflight(ipa_path):
                return False
        else:
            print("⏭️  Skipping TestFlight upload")
            print(f"📦 IPA ready for manual upload: {ipa_path}")

        print("=" * 50)
        print("🎉 Pipeline completed successfully!")
        return True

def load_config(config_path: str) -> Dict:
    """Load configuration from JSON file"""
    try:
        with open(config_path, 'r') as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"❌ Configuration file not found: {config_path}")
        return {}
    except json.JSONDecodeError as e:
        print(f"❌ Invalid JSON in configuration file: {e}")
        return {}

def main():
    parser = argparse.ArgumentParser(description='iOS CI/CD Pipeline for TestFlight')
    parser.add_argument('--config', '-c', default='ios-pipeline-config.json',
                        help='Path to configuration JSON file')
    parser.add_argument('--skip-tests', action='store_true',
                        help='Skip running tests')
    parser.add_argument('--skip-upload', action='store_true',
                        help='Skip uploading to TestFlight')
    parser.add_argument('--scheme', help='Override scheme from config')
    parser.add_argument('--workspace', help='Override workspace from config')
    parser.add_argument('--project', help='Override project from config')

    args = parser.parse_args()

    # Load configuration
    config = load_config(args.config)

    # Override with command line arguments
    if args.scheme:
        config['scheme'] = args.scheme
    if args.workspace:
        config['workspace'] = args.workspace
    if args.project:
        config['project'] = args.project

    # Create and run pipeline
    pipeline = IOSPipeline(config)
    success = pipeline.run_pipeline(
        skip_tests=args.skip_tests,
        skip_upload=args.skip_upload
    )

    sys.exit(0 if success else 1)

if __name__ == '__main__':
    main()

File: testflight/testflight_deployment.ts
import { execSync, ExecSyncOptionsWithStringEncoding } from 'child_process';
import { existsSync, mkdirSync, writeFileSync, readFileSync, readdirSync } from 'fs';
import { join } from 'path';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

interface Config {
    project_path?: string;
    scheme: string;
    workspace?: string;
    project?: string;
    configuration?: string;
    team_id?: string;
    bundle_id?: string;
    app_store_connect_api_key_id?: string;
    app_store_connect_api_issuer_id?: string;
    app_store_connect_api_key_path?: string;
}

class IOSPipeline {
    private config: Config;
    private project_path: string;
    private scheme: string;
    private workspace?: string;
    private project?: string;
    private configuration: string;
    private team_id?: string;
    private bundle_id?: string;
    private app_store_connect_api_key_id?: string;
    private app_store_connect_api_issuer_id?: string;
    private app_store_connect_api_key_path?: string;
    private has_xcpretty: boolean = false;

    constructor(config: Config) {
        this.config = config;
        this.project_path = config.project_path || '.';
        this.scheme = config.scheme;
        this.workspace = config.workspace;
        this.project = config.project;
        this.configuration = config.configuration || 'Release';
        this.team_id = config.team_id;
        this.bundle_id = config.bundle_id;
        this.app_store_connect_api_key_id = config.app_store_connect_api_key_id;
        this.app_store_connect_api_issuer_id = config.app_store_connect_api_issuer_id;
        this.app_store_connect_api_key_path = config.app_store_connect_api_key_path;
    }

    private runCommand(command: string[], cwd: string = this.project_path): string {
        console.log(`🔧 Running: ${command.join(' ')}`);
        const options: ExecSyncOptionsWithStringEncoding = {
            cwd,
            encoding: 'utf-8',
            stdio: ['pipe', 'pipe', 'pipe'],
        };

        try {
            const output = execSync(command.join(' '), options);
            if (output) {
                console.log(`✅ Output: ${output}`);
            }
            return output.toString();
        } catch (error: any) {
            console.error(`❌ Error running command: ${error.message}`);
            console.error(`❌ Stderr: ${error.stderr || ''}`);
            console.error(`❌ Stdout: ${error.stdout || ''}`);
            throw error;
        }
    }

    checkPrerequisites(): boolean {
        console.log('🔍 Checking prerequisites...');

        try {
            this.runCommand(['xcodebuild', '-version']);
        } catch {
            console.error('❌ Xcode is not installed or not in PATH');
            return false;
        }

        try {
            execSync('xcpretty --version', { stdio: 'ignore' });
            this.has_xcpretty = true;
        } catch {
            console.log('⚠️ xcpretty not found. Install with: gem install xcpretty');
            this.has_xcpretty = false;
        }

        if (!this.workspace && !this.project) {
            console.error('❌ Either workspace or project must be specified');
            return false;
        }

        console.log('✅ Prerequisites check passed');
        return true;
    }

    cleanProject(): boolean {
        console.log('🧹 Cleaning project...');
        try {
            const buildTarget = this.workspace
                ? ['-workspace', this.workspace]
                : ['-project', this.project!];
            const command = ['xcodebuild', 'clean', ...buildTarget, '-scheme', this.scheme];
            this.runCommand(command);
            console.log('✅ Project cleaned successfully');
            return true;
        } catch {
            console.error('❌ Failed to clean project');
            return false;
        }
    }

    installDependencies(): boolean {
        const podfilePath = join(this.project_path, 'Podfile');
        if (existsSync(podfilePath)) {
            console.log('📦 Installing CocoaPods dependencies...');
            try {
                this.runCommand(['pod', 'install']);
                console.log('✅ CocoaPods dependencies installed');
                return true;
            } catch {
                console.error('❌ Failed to install CocoaPods dependencies');
                return false;
            }
        } else {
            console.log('ℹ️ No Podfile found, skipping dependency installation');
            return true;
        }
    }

    runTests(): boolean {
        console.log('🧪 Running tests...');
        try {
            const buildTarget = this.workspace
                ? ['-workspace', this.workspace]
                : ['-project', this.project!];
            const command = [
                'xcodebuild',
                'test',
                '-scheme',
                this.scheme,
                '-destination',
                'platform=iOS Simulator,name=iPhone 15,OS=latest',
                ...buildTarget,
            ];

            if (this.has_xcpretty) {
                const xcprettyCommand = [...command, '|', 'xcpretty', '--test', '--color'];
                this.runCommand(xcprettyCommand);
            } else {
                this.runCommand(command);
            }

            console.log('✅ All tests passed');
            return true;
        } catch {
            console.error('❌ Tests failed');
            return false;
        }
    }

    buildArchive(): string | null {
        console.log('🔨 Building and archiving app...');

        const archivePath = join(this.project_path, 'build', `${this.scheme}.xcarchive`);
        mkdirSync(join(this.project_path, 'build'), { recursive: true });

        try {
            const buildTarget = this.workspace
                ? ['-workspace', this.workspace]
                : ['-project', this.project!];
            const command = [
                'xcodebuild',
                'archive',
                '-scheme',
                this.scheme,
                '-configuration',
                this.configuration,
                '-archivePath',
                archivePath,
                '-destination',
                'generic/platform=iOS',
                ...buildTarget,
            ];

            if (this.team_id) {
                command.push('-allowProvisioningUpdates', `DEVELOPMENT_TEAM=${this.team_id}`);
            }

            this.runCommand(command);
            console.log(`✅ Archive created at: ${archivePath}`);
            return archivePath;
        } catch {
            console.error('❌ Failed to build archive');
            return null;
        }
    }

    exportIpa(archivePath: string): string | null {
        console.log('📦 Exporting IPA...');

        const exportPath = join(this.project_path, 'build', 'export');
        mkdirSync(exportPath, { recursive: true });

        const exportOptions: any = {
            method: 'app-store',
            uploadBitcode: false,
            uploadSymbols: true,
            compileBitcode: false,
        };

        if (this.team_id) {
            exportOptions.teamID = this.team_id;
        }

        const exportPlistPath = join(exportPath, 'ExportOptions.plist');
        const plistContent = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>method</key>
  <string>app-store</string>
  <key>uploadBitcode</key>
  <false/>
  <key>uploadSymbols</key>
  <true/>
  <key>compileBitcode</key>
  <false/>
  ${this.team_id ? `<key>teamID</key><string>${this.team_id}</string>` : ''}
</dict>
</plist>`;
        writeFileSync(exportPlistPath, plistContent);

        try {
            const command = [
                'xcodebuild',
                '-exportArchive',
                '-archivePath',
                archivePath,
                '-exportPath',
                exportPath,
                '-exportOptionsPlist',
                exportPlistPath,
            ];

            this.runCommand(command);

            const ipaFiles = readdirSync(exportPath).filter((file: string) => file.endsWith('.ipa'));
            if (ipaFiles.length > 0) {
                const ipaPath = join(exportPath, ipaFiles[0]);
                console.log(`✅ IPA exported to: ${ipaPath}`);
                return ipaPath;
            } else {
                console.error('❌ No IPA file found after export');
                return null;
            }
        } catch {
            console.error('❌ Failed to export IPA');
            return null;
        }
    }

    uploadToTestFlight(ipaPath: string): boolean {
        console.log('🚀 Uploading to TestFlight...');

        try {
            const command = ['xcrun', 'altool', '--upload-app', '--file', ipaPath, '--type', 'ios'];

            if (
                this.app_store_connect_api_key_id &&
                this.app_store_connect_api_issuer_id &&
                this.app_store_connect_api_key_path
            ) {
                command.push(
                    '--apiKey',
                    this.app_store_connect_api_key_id,
                    '--apiIssuer',
                    this.app_store_connect_api_issuer_id
                );
                process.env.API_PRIVATE_KEYS_DIR = join(this.app_store_connect_api_key_path, '..');
            } else {
                console.log('⚠️ App Store Connect API credentials not provided');
                console.log("   You'll need to authenticate manually or provide API credentials");
            }

            this.runCommand(command);
            console.log('✅ Successfully uploaded to TestFlight');
            console.log('ℹ️ Processing may take a few minutes in App Store Connect');
            return true;
        } catch (error: any) {
            console.error(`❌ Upload error: ${error.message}`);
            return false;
        }
    }

    incrementBuildNumber(): boolean {
        console.log('🔢 Incrementing build number...');
        try {
            this.runCommand(['xcrun', 'agvtool', 'next-version', '-all']);
            console.log('✅ Build number incremented');
            return true;
        } catch {
            console.error('❌ Failed to increment build number');
            return false;
        }
    }

    async runPipeline(skipTests: boolean = false, skipUpload: boolean = false): Promise<boolean> {
        console.log('🚀 Starting iOS CI/CD Pipeline');
        console.log('='.repeat(50));

        if (!this.checkPrerequisites()) {
            return false;
        }

        if (!this.installDependencies()) {
            return false;
        }

        if (!this.cleanProject()) {
            return false;
        }

        if (!this.incrementBuildNumber()) {
            console.log('⚠️ Failed to increment build number, continuing...');
        }

        if (!skipTests) {
            if (!this.runTests()) {
                return false;
            }
        } else {
            console.log('⏭️ Skipping tests');
        }

        const archivePath = this.buildArchive();
        if (!archivePath) {
            return false;
        }

        const ipaPath = this.exportIpa(archivePath);
        if (!ipaPath) {
            return false;
        }

        if (!skipUpload) {
            if (!this.uploadToTestFlight(ipaPath)) {
                return false;
            }
        } else {
            console.log('⏭️ Skipping TestFlight upload');
            console.log(`📦 IPA ready for manual upload: ${ipaPath}`);
        }

        console.log('='.repeat(50));
        console.log('🎉 Pipeline completed successfully!');
        return true;
    }
}

function loadConfig(configPath: string): Config {
    try {
        const data = readFileSync(configPath, 'utf-8');
        const config = JSON.parse(data);
        if (!config.scheme) {
            throw new Error('Missing required field: scheme');
        }
        return config;
    } catch (error: any) {
        if (error.code === 'ENOENT') {
            console.error(`❌ Configuration file not found: ${configPath}`);
        } else {
            console.error(`❌ Invalid JSON in configuration file: ${error.message}`);
        }
        return { scheme: '' };
    }
}

async function main(): Promise<void> {
    const argv = yargs(hideBin(process.argv))
        .option('config', { type: 'string', alias: 'c', default: 'ios-pipeline-config.json' })
        .option('scheme', { type: 'string' })
        .option('workspace', { type: 'string' })
        .option('project', { type: 'string' })
        .option('skip-tests', { type: 'boolean', default: false })
        .option('skip-upload', { type: 'boolean', default: false })
        .parseSync();

    const config = loadConfig(argv.config as string);

    // Override config file with CLI args if provided
    if (argv.scheme) config.scheme = argv.scheme;
    if (argv.workspace) config.workspace = argv.workspace;
    if (argv.project) config.project = argv.project;

    const pipeline = new IOSPipeline(config);
    const success = await pipeline.runPipeline(argv['skip-tests'] ?? false, argv['skip-upload'] ?? false);

    process.exit(success ? 0 : 1);
}

main().catch((error) => {
    console.error(`❌ Pipeline failed: ${error.message}`);
    process.exit(1);
});
